### PR TITLE
Aux user object dependency resolving

### LINKED ISSUE #14664: Enhance UO dependency of aux kernels
## Reason
When a user object (UO) name is in the parameter list of an aux kernel, the aux kernel will push the UO into a dependency list (with a overloaded function `getUserObjectTempl`) and later on all the UOs in the list will be put into the `PRE_AUX` for evaluation. However this resolution mechanism is not ideal because the directly dependent UO may depend on other UOs, which will not be pushed into the list and thus later on will not be put into the `PRE_AUX`. Users can of cause add a dummy parameter for the aux kernel which accepts all the indirectly dependent UOs. This is however pretty counter-intuitive and users can often forget and consequently see some strange behavior in the calculation.

## Design
After talk with @permcody , we believe the solution is in `Warehouse`, which has the dependency information that can be used when the code put UOs into `PRE_AUX`. Since the dependency information is not kept currently by `Warehouse`, I will attempt another simpler solution: overload `getUserObjectTempl` in `UserObject`, so that UO itself can know what UOs it depends on. This information can be used in the overloaded function `AuxKernel::getUserObjectTempl`.

## Impact
Make the dependency resolution more complete on the framework side.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #14664.

### CODE REVIEW

**Comment by YaqiWang** [comment]
I have not pushed the last commit in (having it ready now), but want to check if moving few interfaces down to the base class is ok or not first.

**Comment by YaqiWang** [comment]
OK, looks everything is clean. Reviewers, please be noted that the first two commits does not fail any tests. Now I am going to push my last two commits.

File: framework/src/auxkernels/AuxKernel.C [hunk 1]
```diff
@@ -152,25 +152,36 @@ AuxKernelTempl<ComputeValueType>::getSuppliedItems()
...
+AuxKernelTempl<ComputeValueType>::getUserObjectBase(const UserObjectName & name)
 {
-  _depend_uo.insert(_pars.get<UserObjectName>(name));
-  return UserObjectInterface::getUserObjectBase(name);
+  return getUserObjectBaseByName(_pars.get<UserObjectName>(name));
+}
+
+template <typename ComputeValueType>
+const UserObject &
+AuxKernelTempl<ComputeValueType>::getUserObjectBaseByName(const UserObjectName & name)
```
Comment (YaqiWang) [inline framework/src/auxkernels/AuxKernel.C]: This function is missing before which can be really bad.

**Comment by YaqiWang** [comment]
The only arguable code is in the third commit. I suggest to have this in first and then work on the ideal way later on. I personally think that this patches a quite large hole in the framework.

**Comment by YaqiWang** [comment]
@dschwen Possibly the failing phase field test is done by your team, so I asking you to take a look why `grain_volumes` supposes to be a vector postprocessor, but I cannot find any input for adding it?

**Comment by YaqiWang** [comment]
That test could be just a copy-paste error suggested by @dschwen , so we just added a missing vector postprocessor in the input.

**Comment by YaqiWang** [comment]
This is ready for review. Thanks.

**Review by permcody — CHANGES_REQUESTED** [review]
So I think the improvement in functionality is good, but I don't agree with putting a bunch of UserObject and Postprocessor functions staring into the MooseObject class. I'd prefer to see those flow down from the UserObjectInterface and PostprocessorInterface classes as they did before. Was there a problem with that organization?

**Comment by YaqiWang** [comment]
I do not understand the comment. I will stop by sometime.

**Comment by YaqiWang** [comment]
@permcody Just to let you know that I cannot mark those functions with override because they are not virtual in the base class. We basically are not doing overriding but hiding. Same in the `AuxKernel.h`. Unless you want me to make those functions virtual in the base class, we cannot add override here. But I feel it is a rather large change. Theoretically speaking, we should resolve the dependency in warehouse, the codes I did here and the codes in `AuxKernel` are all temporary bandage.

**Review by permcody — APPROVED** [review]

**Pull request merged**